packages/opentelemetry-plugin-http/src/http.ts

```diff
@@ -0,0 +1,413 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BasePlugin, NoopLogger, isValid } from '@opentelemetry/core';
+import {
+  Span,
+  SpanKind,
+  SpanOptions,
+  Logger,
+  Tracer,
+  Attributes,
+} from '@opentelemetry/types';
+import {
+  ClientRequest,
+  IncomingMessage,
+  request,
+  RequestOptions,
+  ServerResponse,
+} from 'http';
+import * as semver from 'semver';
+import * as shimmer from 'shimmer';
+import * as url from 'url';
+import {
+  HttpPluginConfig,
+  Http,
+  Func,
+  ResponseEndArgs,
+  ParsedRequestOptions,
+  HttpRequestArgs,
+} from './types';
+import { Format } from './enums/format';
+import { AttributeNames } from './enums/attributeNames';
+import { Utils } from './utils';
+
+/**
+ * Http instrumentation plugin for Opentelemetry
+ */
+export class HttpPlugin extends BasePlugin<Http> {
+  static readonly component = 'http';
+  options!: HttpPluginConfig;
+  protected _logger!: Logger;
+  protected readonly _tracer!: Tracer;
+
+  constructor(public moduleName: string, public version: string) {
+    super();
+    // TODO: remove this once a logger will be passed
+    // https://github.com/open-telemetry/opentelemetry-js/issues/193
+    this._logger = new NoopLogger();
+  }
+
+  /** Patches HTTP incoming and outcoming request functions. */
+  protected patch() {
+    this._logger.debug(
+      'applying patch to %s@%s',
+      this.moduleName,
+      this.version
+    );
+
+    shimmer.wrap(
```

Comment:
`https.request` forwards to `http.request` therefore this plugin partly handles https also but not all https cases. Not sure how we should cover such inter module side effects.

Reply:
Hi,
From what I know, it doesn't "forwards" but this is 2 separate modules and I didn't notice any issue so far. If I'm wrong please link me the doc. Thanks.

Reply:
It does for Node.js 8 at least (https://github.com/nodejs/node/blob/v8.16.1/lib/https.js#L245)

Reply:
I think we should avoid double tracing as it is confusing. In special because some functions of http module use `http://` in some strings as default which is wrong in that case.

Maybe http module can filter based on `options._defaultAgent === https.globalAgent`?

Another variant would be to combine http and https into a single module as there is quite some reuse of http classes in https anyway. As of now there are not that much conflicts but in case more data should be captured we may end up in more conflicts.

Reply:
I don't think so. The https plugin can't avoid easily that node.js 8 calls `http.request` which will end up in http plugin and a second span will be created.

Reply:
Just remembered that Node 8 goes EOL with end of this year. Maybe just skip this as the problem goes away automatically soon - assuming here that OT has not planned to support EOL versions of node.
Till this happens covering https via http sounds good and easy. If someone complains it can be still improved that time.

Reply:
I think it's planned to keep the support node 8 even after the EOL, few apm vendors (ex: [dynatrace](https://www.dynatrace.com/support/help/technology-support/application-software/nodejs/#support--desupport)) will still support it after that date (see the [discussion about that](https://github.com/open-telemetry/opentelemetry-js/issues/7)). We will need a discussion when it goes EOL to see what to do.

Reply:
Yes, at dynatrace we support EOL versions of NodeJs for 6 more month as far I can remember. But this doesn't imply that OT has to do the same.